src/dataflow/src/render/mod.rs

```diff
@@ -457,95 +414,104 @@ where
-                                    Some(col)
-                                } else {
-                                    None
-                                }
-                            })
-                            .collect::<Vec<_>>();
+                    collection = collection2.as_collection();
+                    err_collection = err_collection.concat(&errors.as_collection());
+                }
 
-                        // Apply predicates and insert dummy values into undemanded columns.
-                        let (collection2, errors) = collection.inner.flat_map_fallible({
-                            let mut datums = crate::render::datum_vec::DatumVec::new();
-                            let mut row_packer = repr::RowPacker::new();
-                            let predicates = std::mem::take(&mut operators.predicates);
-                            // The predicates may be temporal, which requires the nuance
-                            // of an explicit plan capable of evaluating the predicates.
-                            let filter_plan = filter::FilterPlan::create_from(predicates)
-                                .unwrap_or_else(|e| panic!(e));
-                            move |(input_row, time, diff)| {
-                                let mut datums_local = datums.borrow_with(&input_row);
-                                let times_diffs =
-                                    filter_plan.evaluate(&mut datums_local, time, diff);
-                                // The output row may need to have `Datum::Dummy` values stitched in.
-                                let output_row = row_packer.pack(position_or.iter().map(
-                                    |pos_or| match pos_or {
-                                        Some(index) => datums_local[*index],
-                                        None => Datum::Dummy,
-                                    },
-                                ));
-                                // Each produced (time, diff) results in a copy of `output_row` in the output.
-                                // TODO: It would be nice to avoid the `output_row.clone()` for the last output.
-                                times_diffs.map(move |time_diff| {
-                                    time_diff.map(|(t, d)| (output_row.clone(), t, d))
-                                })
-                            }
-                        });
+                // Apply `as_of` to each timestamp.
+                match envelope {
+                    SourceEnvelope::Upsert(_) => {}
+                    _ => {
+                        let as_of_frontier1 = self.as_of_frontier.clone();
+                        collection = collection
+                            .inner
+                            .map_in_place(move |(_, time, _)| {
+                                time.advance_by(as_of_frontier1.borrow())
+                            })
```

Comment:
I'm still reading but it seems like having these before the upsert operator could help, if we believe that it remains correct even with times collapsed down. I think it would if the key used strictly increases, which I think it does for Kafka offsets, but might not for "other things" in the future.

Reply:
Oh wait. Sorry, my code diffing skills failed me and this in the non-upsert brach. Please ignore!